jdbc/src/main/java/org/jclouds/jdbc/service/JdbcService.java

```diff
@@ -16,54 +16,242 @@
+            .findBlobsByDirectory(containerRepository.findContainerByName(containerName), directoryName);
+      result.addAll(blobEntities);
+      if (recursive) {
+         for (BlobEntity blobEntity : blobEntities) {
+            if (blobEntity.isDirectory()) {
+               result.addAll(findBlobsByDirectory(containerName, blobEntity.getKey(), true));
+            }
+         }
+      }
+      return result.build();
+   }
+
+   @Transactional
+   public void deleteBlobsByContainer(String containerName) {
+      List<BlobEntity> blobs = findBlobsByContainer(containerName);
+      for (BlobEntity blob : blobs) {
+         deleteBlob(containerName, blob.getKey());
+      }
+   }
+
+   @Transactional
+   public void deleteBlobsByDirectory(String containerName, String directoryName, boolean recursive) {
+      List<BlobEntity> blobs = findBlobsByDirectory(containerName, directoryName, false);
+      for (BlobEntity blob : blobs) {
+         if (!blob.isDirectory()) {
+            deleteBlob(containerName, blob.getKey());
+         }
+         else {
+            if (recursive) {
+               deleteBlobsByDirectory(containerName, blob.getKey(), true);
+               deleteBlob(containerName, blob.getKey());
+            }
+         }
+      }
+   }
+
+   @Transactional
+   public void deleteBlob(String containerName, String key) {
+      BlobEntity blobEntity = findBlobById(containerName, key);
+      if (blobEntity != null) {
+         deleteChunks(blobEntity.getPayload().getChunks());
+         blobRepository.delete(blobEntity);
+      }
+   }
+
+   @Transactional
+   public void setBlobAccessById(String containerName, String key, BlobAccess access) {
+      BlobEntity blobEntity = findBlobById(containerName, key);
+      blobEntity.setBlobAccess(access);
+      blobRepository.save(blobEntity);
+   }
+
+   @Transactional
+   private void deleteChunks(List<Long> chunkIds) {
+      for (Long chunkId : chunkIds) {
+         chunkRepository.delete(chunkRepository.find(chunkId));
+      }
+   }
+
+   @Transactional(rollbackOn = IOException.class)
+   private List<Long> storeData(InputStream data) throws IOException {
+      ImmutableList.Builder<Long> chunks = ImmutableList.builder();
+      int bytes;
+      byte[] buffer = new byte[JdbcConstants.DEFAULT_CHUNK_SIZE];
+      while ((bytes = data.read(buffer, 0, JdbcConstants.DEFAULT_CHUNK_SIZE)) != -1) {
+         chunks.add(chunkRepository.create(new ChunkEntity(buffer, bytes)).getId());
+         buffer = new byte[JdbcConstants.DEFAULT_CHUNK_SIZE];
+      }
+      data.close();
+      return chunks.build();
+   }
+
+   private void checkIntegrity(Blob blob, BlobEntity entity) throws IOException {
+      HashCode hash = blob.getMetadata().getContentMetadata().getContentMD5AsHashCode();
+      if (hash != null) {
+         HashingInputStream his = new HashingInputStream(Hashing.md5(), new JdbcInputStream(this, entity.getPayload().getChunks()));
```

Comment:
I could make this explicit, but I think it's safe to leave it opened (JdbcInputStream close method does nothing).